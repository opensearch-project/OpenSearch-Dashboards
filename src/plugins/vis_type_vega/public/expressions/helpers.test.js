/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  buildLayerMark,
  buildXAxis,
  buildYAxis,
  cleanString,
  createSpecFromXYChartDatatable,
  formatDatatable,
  setupConfig,
} from './helpers';
import {
  complexDatatable,
  complexVisParams,
  complexDimensions,
  simpleDatatable,
  simpleVisParams,
  simpleDimensions,
  noXAxisDimensions,
} from './__mocks__';
import {
  TEST_DATATABLE_NO_VIS_LAYERS,
  TEST_DATATABLE_NO_VIS_LAYERS_DIRTY,
} from '../../../vis_augmenter/public';

describe('helpers', function () {
  describe('formatDatatable()', function () {
    it('formatSimpleDatatable', function () {
      expect(formatDatatable(TEST_DATATABLE_NO_VIS_LAYERS)).toBe(TEST_DATATABLE_NO_VIS_LAYERS);
    });
    it('formatDirtyDatatable', function () {
      expect(formatDatatable(TEST_DATATABLE_NO_VIS_LAYERS_DIRTY)).toStrictEqual(
        TEST_DATATABLE_NO_VIS_LAYERS
      );
    });
  });

  describe('cleanString()', function () {
    it('string should not contain quotation marks', function () {
      const dirtyString = '"someString"';
      expect(cleanString(dirtyString)).toBe('someString');
    });
  });

  describe('setupConfig()', function () {
    it('check all legend positions', function () {
      const visAugmenterConfig = {
        some: 'config',
      };
      const baseConfig = {
        view: {
          stroke: null,
        },
        concat: {
          spacing: 0,
        },
        legend: {
          orient: null,
        },
        kibana: {
          hideWarnings: true,
          visAugmenterConfig,
        },
      };
      const positions = ['top', 'right', 'left', 'bottom'];
      positions.forEach((position) => {
        const visParams = { legendPosition: position };
        baseConfig.legend.orient = position;
        baseConfig.legend.offset = position === 'top' || position === 'bottom' ? 0 : 18;
        expect(setupConfig(visParams, visAugmenterConfig)).toStrictEqual(baseConfig);
      });
    });
  });

  describe('buildLayerMark()', function () {
    const types = ['line', 'area', 'histogram'];
    const interpolates = ['linear', 'cardinal', 'step-after'];
    const strokeWidths = [-1, 0, 1, 2, 3, 4];
    const showCircles = [false, true];

    it('check each mark possible value', function () {
      const mark = {
        type: null,
        interpolate: null,
        strokeWidth: null,
        point: null,
      };
      types.forEach((type) => {
        mark.type = type;
        interpolates.forEach((interpolate) => {
          mark.interpolate = interpolate;
          strokeWidths.forEach((strokeWidth) => {
            mark.strokeWidth = strokeWidth;
            showCircles.forEach((showCircle) => {
              mark.point = showCircle;
              const param = {
                type: type,
                interpolate: interpolate,
                lineWidth: strokeWidth,
                showCircles: showCircle,
              };
              expect(buildLayerMark(param)).toStrictEqual(mark);
            });
          });
        });
      });
    });
  });

  describe('buildXAxis()', function () {
    it('build different XAxis', function () {
      const xAxisTitle = 'someTitle';
      const xAxisId = 'someId';
      [true, false].forEach((enableGrid) => {
        const visParams = { grid: { categoryLines: enableGrid } };
        const vegaXAxis = {
          axis: {
            title: xAxisTitle,
            grid: enableGrid,
          },
          field: xAxisId,
          type: 'temporal',
        };
        expect(buildXAxis(xAxisTitle, xAxisId, visParams)).toStrictEqual(vegaXAxis);
      });
    });
  });

  describe('buildYAxis()', function () {
    it('build different YAxis', function () {
      const valueAxis = {
        id: 'someId',
        labels: {
          rotate: 75,
          show: false,
        },
        position: 'left',
        title: {
          text: 'someText',
        },
      };
      const column = { name: 'columnName', id: 'columnId' };
      const visParams = { grid: { valueAxis: true } };
      const vegaYAxis = {
        axis: {
          title: 'someText',
          grid: true,
          orient: 'left',
          labels: false,
          labelAngle: 75,
        },
        field: 'columnId',
        type: 'quantitative',
      };
      expect(buildYAxis(column, valueAxis, visParams)).toStrictEqual(vegaYAxis);

      valueAxis.title.text = '""';
      vegaYAxis.axis.title = 'columnName';
      expect(buildYAxis(column, valueAxis, visParams)).toStrictEqual(vegaYAxis);
    });
    it('build YAxis with percentile rank', function () {
      const valueAxis = {
        id: 'someId',
        labels: {
          rotate: 75,
          show: false,
        },
        position: 'left',
        title: {
          text: 'someText',
        },
      };
      const column = { name: 'columnName', id: 'columnId', meta: { type: 'percentile_ranks' } };
      const visParams = { grid: { valueAxis: true } };
      const vegaYAxis = {
        axis: {
          title: 'someText',
          grid: true,
          orient: 'left',
          labels: false,
          labelAngle: 75,
          format: '.0%',
        },
        field: 'columnId',
        type: 'quantitative',
      };
      expect(buildYAxis(column, valueAxis, visParams)).toStrictEqual(vegaYAxis);
    });
  });

  describe('createSpecFromXYChartDatatable()', function () {
    // Following 3 tests fail since they are persisting temporal data
    // which can cause snapshots to fail depending on the test env they are run on.
    // Tracking issue: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4555
    // TODO: Add a test for the fix in this PR: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4554
    it.skip('build simple line chart"', function () {
      expect(
        JSON.stringify(
          createSpecFromXYChartDatatable(
            formatDatatable(JSON.parse(simpleDatatable)),
            JSON.parse(simpleVisParams),
            JSON.parse(simpleDimensions)
          )
        )
      ).toMatchSnapshot();
    });

    it.skip('build empty chart if no x-axis is defined"', function () {
      expect(
        JSON.stringify(
          createSpecFromXYChartDatatable(
            formatDatatable(JSON.parse(simpleDatatable)),
            JSON.parse(simpleVisParams),
            JSON.parse(noXAxisDimensions)
          )
        )
      ).toMatchSnapshot();
    });

    it.skip('build complicated line chart"', function () {
      expect(
        JSON.stringify(
          createSpecFromXYChartDatatable(
            formatDatatable(JSON.parse(complexDatatable)),
            JSON.parse(complexVisParams),
            JSON.parse(complexDimensions)
          )
        )
      ).toMatchSnapshot();
    });
  });
});
