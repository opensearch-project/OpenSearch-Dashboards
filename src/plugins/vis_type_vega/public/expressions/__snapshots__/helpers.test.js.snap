// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers createSpecFromXYChartDatatable() build complicated line chart" 1`] = `"{\\"$schema\\":\\"https://vega.github.io/schema/vega-lite/v5.json\\",\\"data\\":{\\"values\\":[{\\"col-0-2\\":1672214400000,\\"col-1-1\\":44,\\"col-2-3\\":60.9375},{\\"col-0-2\\":1672300800000,\\"col-1-1\\":150,\\"col-2-3\\":82.5},{\\"col-0-2\\":1672387200000,\\"col-1-1\\":154,\\"col-2-3\\":79.5},{\\"col-0-2\\":1672473600000,\\"col-1-1\\":144,\\"col-2-3\\":75.875},{\\"col-0-2\\":1672560000000,\\"col-1-1\\":133,\\"col-2-3\\":259.25},{\\"col-0-2\\":1672646400000,\\"col-1-1\\":149,\\"col-2-3\\":90},{\\"col-0-2\\":1672732800000,\\"col-1-1\\":152,\\"col-2-3\\":79.0625},{\\"col-0-2\\":1672819200000,\\"col-1-1\\":144,\\"col-2-3\\":82.5},{\\"col-0-2\\":1672905600000,\\"col-1-1\\":166,\\"col-2-3\\":85.25},{\\"col-0-2\\":1672992000000,\\"col-1-1\\":151,\\"col-2-3\\":92},{\\"col-0-2\\":1673078400000,\\"col-1-1\\":143,\\"col-2-3\\":90.75},{\\"col-0-2\\":1673164800000,\\"col-1-1\\":148,\\"col-2-3\\":92},{\\"col-0-2\\":1673251200000,\\"col-1-1\\":146,\\"col-2-3\\":83.25},{\\"col-0-2\\":1673337600000,\\"col-1-1\\":137,\\"col-2-3\\":98},{\\"col-0-2\\":1673424000000,\\"col-1-1\\":152,\\"col-2-3\\":83.6875},{\\"col-0-2\\":1673510400000,\\"col-1-1\\":152,\\"col-2-3\\":83.6875},{\\"col-0-2\\":1673596800000,\\"col-1-1\\":151,\\"col-2-3\\":87.4375},{\\"col-0-2\\":1673683200000,\\"col-1-1\\":157,\\"col-2-3\\":63.75},{\\"col-0-2\\":1673769600000,\\"col-1-1\\":151,\\"col-2-3\\":81.5625},{\\"col-0-2\\":1673856000000,\\"col-1-1\\":152,\\"col-2-3\\":100.6875},{\\"col-0-2\\":1673942400000,\\"col-1-1\\":142,\\"col-2-3\\":98},{\\"col-0-2\\":1674028800000,\\"col-1-1\\":151,\\"col-2-3\\":100.8125},{\\"col-0-2\\":1674115200000,\\"col-1-1\\":163,\\"col-2-3\\":83.6875},{\\"col-0-2\\":1674201600000,\\"col-1-1\\":156,\\"col-2-3\\":85.8125},{\\"col-0-2\\":1674288000000,\\"col-1-1\\":153,\\"col-2-3\\":98},{\\"col-0-2\\":1674374400000,\\"col-1-1\\":162,\\"col-2-3\\":75.9375},{\\"col-0-2\\":1674460800000,\\"col-1-1\\":152,\\"col-2-3\\":113.375},{\\"col-0-2\\":1674547200000,\\"col-1-1\\":159,\\"col-2-3\\":73.625},{\\"col-0-2\\":1674633600000,\\"col-1-1\\":165,\\"col-2-3\\":72.8125},{\\"col-0-2\\":1674720000000,\\"col-1-1\\":153,\\"col-2-3\\":113.375},{\\"col-0-2\\":1674806400000,\\"col-1-1\\":149,\\"col-2-3\\":82.5},{\\"col-0-2\\":1674892800000,\\"col-1-1\\":94,\\"col-2-3\\":54}]},\\"config\\":{\\"view\\":{\\"stroke\\":null},\\"concat\\":{\\"spacing\\":0},\\"legend\\":{\\"orient\\":\\"bottom\\"},\\"kibana\\":{\\"hideWarnings\\":true}},\\"layer\\":[{\\"mark\\":{\\"type\\":\\"line\\",\\"interpolate\\":\\"linear\\",\\"strokeWidth\\":2,\\"point\\":true},\\"encoding\\":{\\"x\\":{\\"axis\\":{\\"title\\":\\"order_date per day\\",\\"grid\\":false},\\"field\\":\\"col-0-2\\",\\"type\\":\\"temporal\\"},\\"y\\":{\\"axis\\":{\\"title\\":\\"Count\\",\\"grid\\":true,\\"orient\\":\\"right\\",\\"labels\\":true,\\"labelAngle\\":75},\\"field\\":\\"col-1-1\\",\\"type\\":\\"quantitative\\"},\\"tooltip\\":[{\\"field\\":\\"col-0-2\\",\\"type\\":\\"temporal\\",\\"title\\":\\"order_date per day\\"},{\\"field\\":\\"col-1-1\\",\\"type\\":\\"quantitative\\",\\"title\\":\\"Count\\"}],\\"color\\":{\\"datum\\":\\"Count\\"}}},{\\"mark\\":{\\"type\\":\\"line\\",\\"interpolate\\":\\"linear\\",\\"strokeWidth\\":2,\\"point\\":true},\\"encoding\\":{\\"x\\":{\\"axis\\":{\\"title\\":\\"order_date per day\\",\\"grid\\":false},\\"field\\":\\"col-0-2\\",\\"type\\":\\"temporal\\"},\\"y\\":{\\"axis\\":{\\"title\\":\\"Count\\",\\"grid\\":true,\\"orient\\":\\"right\\",\\"labels\\":true,\\"labelAngle\\":75},\\"field\\":\\"col-2-3\\",\\"type\\":\\"quantitative\\"},\\"tooltip\\":[{\\"field\\":\\"col-0-2\\",\\"type\\":\\"temporal\\",\\"title\\":\\"order_date per day\\"},{\\"field\\":\\"col-2-3\\",\\"type\\":\\"quantitative\\",\\"title\\":\\"Max products.min_price\\"}],\\"color\\":{\\"datum\\":\\"Max products.min_price\\"}}},{\\"mark\\":\\"rule\\",\\"encoding\\":{\\"x\\":{\\"type\\":\\"temporal\\",\\"field\\":\\"now_field\\"},\\"color\\":{\\"value\\":\\"red\\"},\\"size\\":{\\"value\\":1}}},{\\"mark\\":{\\"type\\":\\"rule\\",\\"color\\":\\"#E7664C\\",\\"strokeDash\\":[8,8]},\\"encoding\\":{\\"y\\":{\\"datum\\":100}}}],\\"transform\\":[{\\"calculate\\":\\"now()\\",\\"as\\":\\"now_field\\"}]}"`;

exports[`helpers createSpecFromXYChartDatatable() build empty chart if no x-axis is defined" 1`] = `"{\\"$schema\\":\\"https://vega.github.io/schema/vega-lite/v5.json\\",\\"data\\":{\\"values\\":[{\\"col-0-2\\":1672214400000,\\"col-1-1\\":44},{\\"col-0-2\\":1672300800000,\\"col-1-1\\":150},{\\"col-0-2\\":1672387200000,\\"col-1-1\\":154},{\\"col-0-2\\":1672473600000,\\"col-1-1\\":144},{\\"col-0-2\\":1672560000000,\\"col-1-1\\":133},{\\"col-0-2\\":1672646400000,\\"col-1-1\\":149},{\\"col-0-2\\":1672732800000,\\"col-1-1\\":152},{\\"col-0-2\\":1672819200000,\\"col-1-1\\":144},{\\"col-0-2\\":1672905600000,\\"col-1-1\\":166},{\\"col-0-2\\":1672992000000,\\"col-1-1\\":151},{\\"col-0-2\\":1673078400000,\\"col-1-1\\":143},{\\"col-0-2\\":1673164800000,\\"col-1-1\\":148},{\\"col-0-2\\":1673251200000,\\"col-1-1\\":146},{\\"col-0-2\\":1673337600000,\\"col-1-1\\":137},{\\"col-0-2\\":1673424000000,\\"col-1-1\\":152},{\\"col-0-2\\":1673510400000,\\"col-1-1\\":152},{\\"col-0-2\\":1673596800000,\\"col-1-1\\":151},{\\"col-0-2\\":1673683200000,\\"col-1-1\\":157},{\\"col-0-2\\":1673769600000,\\"col-1-1\\":151},{\\"col-0-2\\":1673856000000,\\"col-1-1\\":152},{\\"col-0-2\\":1673942400000,\\"col-1-1\\":142},{\\"col-0-2\\":1674028800000,\\"col-1-1\\":151},{\\"col-0-2\\":1674115200000,\\"col-1-1\\":163},{\\"col-0-2\\":1674201600000,\\"col-1-1\\":156},{\\"col-0-2\\":1674288000000,\\"col-1-1\\":153},{\\"col-0-2\\":1674374400000,\\"col-1-1\\":162},{\\"col-0-2\\":1674460800000,\\"col-1-1\\":152},{\\"col-0-2\\":1674547200000,\\"col-1-1\\":159},{\\"col-0-2\\":1674633600000,\\"col-1-1\\":165},{\\"col-0-2\\":1674720000000,\\"col-1-1\\":153},{\\"col-0-2\\":1674806400000,\\"col-1-1\\":149},{\\"col-0-2\\":1674892800000,\\"col-1-1\\":94}]},\\"config\\":{\\"view\\":{\\"stroke\\":null},\\"concat\\":{\\"spacing\\":0},\\"legend\\":{\\"orient\\":\\"right\\"},\\"kibana\\":{\\"hideWarnings\\":true}},\\"layer\\":[]}"`;

exports[`helpers createSpecFromXYChartDatatable() build simple line chart" 1`] = `"{\\"$schema\\":\\"https://vega.github.io/schema/vega-lite/v5.json\\",\\"data\\":{\\"values\\":[{\\"col-0-2\\":1672214400000,\\"col-1-1\\":44},{\\"col-0-2\\":1672300800000,\\"col-1-1\\":150},{\\"col-0-2\\":1672387200000,\\"col-1-1\\":154},{\\"col-0-2\\":1672473600000,\\"col-1-1\\":144},{\\"col-0-2\\":1672560000000,\\"col-1-1\\":133},{\\"col-0-2\\":1672646400000,\\"col-1-1\\":149},{\\"col-0-2\\":1672732800000,\\"col-1-1\\":152},{\\"col-0-2\\":1672819200000,\\"col-1-1\\":144},{\\"col-0-2\\":1672905600000,\\"col-1-1\\":166},{\\"col-0-2\\":1672992000000,\\"col-1-1\\":151},{\\"col-0-2\\":1673078400000,\\"col-1-1\\":143},{\\"col-0-2\\":1673164800000,\\"col-1-1\\":148},{\\"col-0-2\\":1673251200000,\\"col-1-1\\":146},{\\"col-0-2\\":1673337600000,\\"col-1-1\\":137},{\\"col-0-2\\":1673424000000,\\"col-1-1\\":152},{\\"col-0-2\\":1673510400000,\\"col-1-1\\":152},{\\"col-0-2\\":1673596800000,\\"col-1-1\\":151},{\\"col-0-2\\":1673683200000,\\"col-1-1\\":157},{\\"col-0-2\\":1673769600000,\\"col-1-1\\":151},{\\"col-0-2\\":1673856000000,\\"col-1-1\\":152},{\\"col-0-2\\":1673942400000,\\"col-1-1\\":142},{\\"col-0-2\\":1674028800000,\\"col-1-1\\":151},{\\"col-0-2\\":1674115200000,\\"col-1-1\\":163},{\\"col-0-2\\":1674201600000,\\"col-1-1\\":156},{\\"col-0-2\\":1674288000000,\\"col-1-1\\":153},{\\"col-0-2\\":1674374400000,\\"col-1-1\\":162},{\\"col-0-2\\":1674460800000,\\"col-1-1\\":152},{\\"col-0-2\\":1674547200000,\\"col-1-1\\":159},{\\"col-0-2\\":1674633600000,\\"col-1-1\\":165},{\\"col-0-2\\":1674720000000,\\"col-1-1\\":153},{\\"col-0-2\\":1674806400000,\\"col-1-1\\":149},{\\"col-0-2\\":1674892800000,\\"col-1-1\\":94}]},\\"config\\":{\\"view\\":{\\"stroke\\":null},\\"concat\\":{\\"spacing\\":0},\\"legend\\":{\\"orient\\":\\"right\\"},\\"kibana\\":{\\"hideWarnings\\":true}},\\"layer\\":[{\\"mark\\":{\\"type\\":\\"line\\",\\"interpolate\\":\\"linear\\",\\"strokeWidth\\":2,\\"point\\":true},\\"encoding\\":{\\"x\\":{\\"axis\\":{\\"title\\":\\"order_date per day\\",\\"grid\\":false},\\"field\\":\\"col-0-2\\",\\"type\\":\\"temporal\\"},\\"y\\":{\\"axis\\":{\\"title\\":\\"Count\\",\\"grid\\":false,\\"orient\\":\\"left\\",\\"labels\\":true,\\"labelAngle\\":0},\\"field\\":\\"col-1-1\\",\\"type\\":\\"quantitative\\"},\\"tooltip\\":[{\\"field\\":\\"col-0-2\\",\\"type\\":\\"temporal\\",\\"title\\":\\"order_date per day\\"},{\\"field\\":\\"col-1-1\\",\\"type\\":\\"quantitative\\",\\"title\\":\\"Count\\"}],\\"color\\":{\\"datum\\":\\"Count\\"}}}]}"`;
